include/flang/Semantics/symbol.h

```diff
@@ -131,8 +131,18 @@ class AssocEntityDetails : public EntityDetails {
   AssocEntityDetails &operator=(AssocEntityDetails &&) = default;
   const MaybeExpr &expr() const { return expr_; }
 
+  void set_rank(int rank);
+  std::optional<int> associationRank() const {
+    if (associationRank_.has_value()) {
+      return associationRank_.value();
+    } else {
+      return {};
+    }
+  }
+
 private:
   MaybeExpr expr_;
+  std::optional<int> associationRank_;
```

Comment:
Is this new field relevant to the shape analysis code in `lib/Evaluate/shape.cpp`?  I think that it must be, so that code should be extended to use the specific case ranks so that operations that depend on the shape of an array will be compiled correctly.

Reply:
### Can you please explain more on what do you mean by 

> new field relevant to the shape analysis code


The tests other than select rank which depend on shape of array seem to be successful for  `ninja test check-all` 

What I can understand from comments is that, 
` MaybeExpr expr_` private member function in `class AssocEntityDetails` is being used in `lib/Evaluate/shape.cpp` inside function `GetShape() `
and my patch doesn't seem to handle the `expr_` inside `SetRankFromParserNode `but is handled in `SetTypeFromAssociation`

Reply:
Make sure that your tests for `SELECT RANK` include operations in the rank cases that will force the compiler to analyze the shape of the array at the specific rank.  Use `FORALL` statements, the `SHAPE` intrinsic, array assignments, &c.  There is probably code in the shape analysis framework that will need to be updated to handle this new representation of construct entities correctly, using the rank of the construct entity and the descriptor of the assumed-rank dummy argument for inquiries.